cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/store/AppMetadataStore.java

```diff
@@ -569,6 +569,57 @@ public RunRecordMeta recordProgramDeprovisioned(ProgramRunId programRunId, byte[
     return meta;
   }
 
+  /**
+   * Record that the program run has been orphaned. If the current status has a higher source id,
+   * this call will be ignored.
+   *
+   * @param programRunId program run
+   * @param sourceId unique id representing the source of program run status, such as the message id of the program
+   *                 run status notification in TMS. The source id must increase as the recording time of the program
+   *                 run status increases, so that the attempt to persist program run status older than the existing
+   *                 program run status will be ignored
+   * @return {@link RunRecordMeta} that was persisted, or {@code null} if the update was ignored.
+   */
+  @Nullable
+  public RunRecordMeta recordProgramOrphaned(ProgramRunId programRunId, byte[] sourceId) {
+    MDSKey.Builder key = getProgramKeyBuilder(TYPE_RUN_RECORD_COMPLETED, programRunId.getParent());
+
+    RunRecordMeta existing = getRun(programRunId);
+    if (existing == null) {
+      LOG.debug("Ignoring unexpected transition of program run {} to cluster state {} with no existing run record.",
+                programRunId, ProgramRunClusterStatus.DEPROVISIONED);
+      return null;
+    } else if (!isValid(existing, sourceId, "orphaned")) {
+      return null;
+    }
+
+    ProgramRunClusterStatus clusterStatus = existing.getCluster().getStatus();
+    // should only be able to get to orphaned from the provisioning or deprovisioning states
+    if (clusterStatus != ProgramRunClusterStatus.PROVISIONING &&
```

Comment:
One suggestion is that the logic can be encapsulated in the enum class itself such that in here it becomes

```java
if (!clusterStatus.canTransitTo(ProgramRunClusterStatus.ORPHANED)) {
  LOG.warn(...);
  return null;
}
```

and it can applies to all other states transitions.